### DISCUSSION TITLE
Merge two images

### DISCUSSION THREAD

**Discussion by Schrank97** [ImageMagick/ImageMagick#3299] [Help]
Hello Guys,

I'm trying to merge two images to one image. Here the specs for the two images:
ColorSpace= Both: CMYK
FileFormat = Both: .Tiff
Implemented Colors: One has all colors, the other one includes only Black (K-Channel) and white pixels.

I'm using the .Net-DLLs.
The one with all Colors is the Base-Image. I want all Black Pixels from the other image to be in the Colored one.
I was trying to look at each pixel, and if the K-Channel is used, i copied the Pixel and applied the value to the Colored Image. I did this promatically. And that needs a long time to execute this function.

Then i tried the "Composite-Function." The operator "CopyBlack" is nearly that, what I've been looking for. But he deletes the Black Pixels from the Colored- Image. That shouldn't happen.
The "Overlay"-Operator didn't work neither, because the images, like a said, are cmyk and have no transparency. This takes the White Pixels from the Black-Image and.... the colored Image is gone. Converting the images to PNG, Apply Transparency, Combine them, and finally Convert them back to tiff -> Needs to much time for execution.

Is there a better way to achieve the "Both in One"-Solution ? Did i miss a Useful Composite- Operator?
Execution time does really matter for me on this problem, so i would be grateful for every help! :)

Best Regards
Jakob

![Image](https://user-images.githubusercontent.com/75781929/109162359-676e6100-7778-11eb-9266-d98085d81993.png)

**Comment by Chetan-Goyal** [comment]
I think [composite](https://imagemagick.org/script/composite.php) can be helpful in your case.

**Reply by Schrank97** [reply]
Hello Chetan-Goyal,

thank you for reply. But like i mentioned in the Question above, i already tried the Composite-function. With many Operators. But every operator i used was not giving the expected result.
"CopyBlack" was nearly right, but it deleted all Black-Channels from the Colored Profile, which made the image quite strange.

**Reply by Chetan-Goyal** [reply]
You can remove background from black and white image and then you can overlap the two images. Let me know if this works.

**Reply by Schrank97** [reply]
Already tried that before. But CMYK-Images don't let transparency be inside. So i can't remove the background. The Pixels there will get the Values 0,0,0,K. If i composite them then, it overwrites every pixel with the 0-Values and the image is getting overwritten completely.

**Reply by Chetan-Goyal** [reply]
I guess I will not be able to help you. I will let you know if I find any other way which could work.

**Reply by Schrank97** [reply]
Thank you for your help anyways! :)

**Comment by snibgo** [comment]
I don't understand what you want.

I suggest you link to two sample inputs, and the required output.

**Reply by Schrank97** [reply]
Hey,

i tried to upload the images before, but the forum doesn't allow .tif- files.
I converted the images to jpgs, so you can see, what I'm "thinking of". The base images are .tiff-Files in CMYK-Space

Also there's an image added, which shows, what happens, if i do the Composite-Function with CopyBlack-Operator.


![BlackImage](https://user-images.githubusercontent.com/75781929/109300021-48360900-7836-11eb-9bf8-add4643ff2ea.jpg)
![ColoredImage](https://user-images.githubusercontent.com/75781929/109300027-4a986300-7836-11eb-9542-bda7ff0b9d5b.jpg)
![ExpectedResult](https://user-images.githubusercontent.com/75781929/109300033-4d935380-7836-11eb-9593-28b63d44d013.jpg)

(and here, how it should NOT look like, but looks like right now)
![Result_CompositeCopyBlack](https://user-images.githubusercontent.com/75781929/109300052-54ba6180-7836-11eb-8379-16455490dbd3.jpg)

Thank you for looking at it, Snibgo :)

**Comment by snibgo** [comment]
Is this what you want? Windows BAT script.
```
magick ^
  109300021-48360900-7836-11eb-9bf8-add4643ff2ea.jpg ^
  -colorspace sRGB ^
  ( 109300027-4a986300-7836-11eb-9542-bda7ff0b9d5b.jpg ^
    -colorspace sRGB ^
    -transparent White ^
  ) ^
  -composite ^
  x.png
```
![x](https://user-images.githubusercontent.com/33812368/109303482-c7c1d900-7832-11eb-8ef0-885a5f1b4a49.png)

If you want a CMYK output, then:
```
magick ^
  109300021-48360900-7836-11eb-9bf8-add4643ff2ea.jpg ^
  -colorspace sRGB ^
  ( 109300027-4a986300-7836-11eb-9542-bda7ff0b9d5b.jpg ^
    -colorspace sRGB ^
    -transparent White ^
  ) ^
  -composite ^
  -colorspace CMYK ^
  x.tiff
```

**Reply by Schrank97** [reply]
Yes, that's totally the right output. I was just wondering, if it is possible to do it faster. Time does really matter in this project. 

Thanks for your help on this one! :)
Maybe i can safe some time on an other step.

Best regards
Jakob
